### PR TITLE
Ported AndroidAsync Style Socket.IO API to standalone library

### PR DESCRIPTION
**Pull request by vinaysshenoy**
I have a few doubts on how to handle certain things - 
1. Earlier your AsyncServer would take care of queueing multiple SocketIO connect requests(for the initial setup). I'm currently using an AsyncTask that runs on a single thread. I'm thinking to have a Queue and the WebSocket's connect method runs a loop through the queue to connect websockets. We can make a relatively efficient queueing mechanism this way.
2. Is it worth it to switch between HttpClient and HttpUrlConnection for the webosocket handshake for device OS's? I thought of implementing it, but I felt it's not worth based on how much they are used.
3. The EventCallback method should be sending the event names as well. This way we can have a single listener object that listens for events. Or even route groups of events to different listeners.

### CODE REVIEW

**Comment by koush** [comment]
Looks good, though I'm not a huge fan of changing WebSocketClient to WebSocket, and changing the Handler to distinct Callbacks. There's no need to break backwards compatibility there. Was here a reason for doing that?

I'd rather the corresponding changes were made in SocketIOClient to support WebSocketClient vs WebSocket. Should be relatively straightforward.

**Comment by koush** [comment]
Your questions:
1) If you use a send handler as in the original implementation, this would take care of queueing, wouldn't it?
2) Doesn't matter IMO.
3) Good point, please edit that and submit a fix to AndroidAsync as well if you want :)

**Comment by vinaysshenoy** [comment]
Oh, I thought we looking to move completely away from the old library style. I'll revert some things back to the older style.

I'll submit a fix for AndroidAsync as well. Btw, Are you planning to upgrade the README for EventCallbacks? It took me quite a while to figure out that I should be using client.addListener() instead of client.setEventCallback() which seemed to be missing.. I can submit a fix for that too.. but it might take a while as I'll do these only after completing the changes to this library.

**Comment by koush** [comment]
Oh, I completely forgot about the README. I had just updated the tests. I can fix that.

**Comment by vinaysshenoy** [comment]
I'm working on adding the earlier Handler style listener back to SocketIOClient. Would you mind if I renamed the interface to SocketIOCallbacks or SocketIOHandler instead of Handler?

**Comment by koush** [comment]
Go for it

**Comment by vinaysshenoy** [comment]
All right. And one more thing, in AndroidAsync, you can subscribe to events, while in the earlier android-websockets, you'd receive all events. Which one would you prefer to bring into this?

If we keep the new version(which I personally prefer), then in the onConnectCompleted(), we'll need to call

``` java
client.setSocketIOCallbacks(new SocketIOCallbacks());
client.listenForEvents(Arrays.asList("news", "chat"));
```

The final event will still come though `Handler#on(event, args)`.

**Comment by koush** [comment]
Sorry, I was confused earlier. I thought you meant android.os.Handler.

Right now, none of the events fire back onto the main thread, and I thought you were fixing that.

The socket.io callbacks should remain similar to the ones exposed by AndroidAsync. Separate callbacks for Strings, JSON, and Event. And event callbacks are exposed via EventEmitter.

What are you trying to solve?

**Comment by koush** [comment]
To clarify, I do not want the old style SocketIOClient.Handler. In retrospect, I found that implementation kludgy.

However, I do think that events should be invoked onto the main ui thread. I'll handle that, as I need to do similar for AndroidAsync as well.

**Comment by vinaysshenoy** [comment]
I got confused too.. I thought that you preferred to expose socket messages via the earlier SocketIO.Handler interface.. So I added that in. And I wanted to rename it to SocketIOCallbacks since it was having the same name as an Android system class. Anyway, I'll revert it back.. Won't be too much of a problem.

**Comment by koush** [comment]
Cool, sorry for the confusion. Should have read your comment more thoroughly.

I suppose for the purposes of your existing code, you'd want to keep a SocketIOHandler around for easy transition. I think that can just be done outside of the library just as well though.

**Comment by vinaysshenoy** [comment]
Not a problem. I've already removed the interface.

As for the callbacks, I'm adding code to invoke them on the main thread now. Or rather, onto the same Handler that is passed to SocketIOClient.

**Comment by vinaysshenoy** [comment]
All right. It should be complete now. Could you review and let me know if I need to make any changes?

**Comment by koush** [comment]
There are changes to WebSocketClient and HybiParser. Member names, package renames, formatting, whitespace, etc. Those shouldn't be in the scope of this change.

The websocket/hybiparser should remain as untouched as possible.

**Comment by koush** [comment]
Those are originally created by @codebutler, so I want them to remain in his package prefix.

**Comment by vinaysshenoy** [comment]
Right, I'll fix those.

**Comment by vinaysshenoy** [comment]
HybiParser and WebSocketClient are mostly the same as earlier(apart from the package, which I'll be changing).

The parts I did change corresponds to the WebSocketClient#Listener interface  to conform to the callback methods in the attach() method in SocketIOConnection. Should I revert those as well?

**Comment by koush** [comment]
WebSocket is still using separate callbacks instead of Listener. I'll leave comments inline

File: src/com/codebutler/android_websockets/HybiParser.java [hunk 1]
```diff
@@ -30,16 +30,20 @@
...
+import java.io.UnsupportedEncodingException;
 import java.util.Arrays;
 import java.util.List;
 
+import android.util.Log;
+
 public class HybiParser {
     private static final String TAG = "HybiParser";
 
-    private WebSocketClient mClient;
```
Comment (koush) [inline src/com/codebutler/android_websockets/HybiParser.java:42]: This rename is unnecessary, let's leave that alone.

If you want to change WebSocket internals around, we should do that in a separate pull request.

File: src/com/codebutler/android_websockets/WebSocketClient.java [hunk 2]
```diff
@@ -233,12 +256,47 @@ public void run() {
         });
     }
 
-    public interface Listener {
-        public void onConnect();
-        public void onMessage(String message);
-        public void onMessage(byte[] data);
-        public void onDisconnect(int code, String reason);
-        public void onError(Exception error);
+    public static interface Listener {
```
Comment (koush) [inline src/com/codebutler/android_websockets/WebSocketClient.java]: Let's continue using a Listener for all callbacks on WebSocketClient. Let's not change the WebSocket implementation.

I want to retain backwards WebSocket functionality, as it is not broken or in need of love. It's only the socket.io classes that are in need of refactoring.

**Comment by koush** [comment]
Cool, almost there, just need to go back to the original websocket implementation (as much as is reasonably possible) and we can accept.

**Comment by codebutler** [comment]
FYI I have some patches from someone that cleans up the code a bit and removes the android dependency. https://github.com/codebutler/android-websockets/issues/10 I'd eventually like to get this thing packaged into a proper library (as tinywebsockets or something not-android-specific) and posted on maven central so projects like this could easily depend on it.

File: src/com/codebutler/android_websockets/WebSocketClient.java [hunk 3]
```diff
@@ -25,18 +10,47 @@
...
+import android.os.Handler;
+import android.os.HandlerThread;
+import android.text.TextUtils;
+import android.util.Base64;
+import android.util.Log;
+
 public class WebSocketClient {
     private static final String TAG = "WebSocketClient";
 
-    private URI                      mURI;
```
Comment (koush) [inline src/com/codebutler/android_websockets/WebSocketClient.java:31]: Though I'm not a fan of the tab/spacing here, let's not change it (or do it in a separate patch). Future merges with codebutler's fork would be a pain.

Try to avoid whitespace changes in general, or put them in a separate pull request so they can be reviewed as they are separate issues. Commits and pull requests should tackle one problem at a time, rather than bunch up a number of different changes.

File: src/com/codebutler/android_websockets/WebSocketClient.java [hunk 4]
```diff
@@ -25,18 +10,47 @@
...
-    private HybiParser               mParser;
-    private boolean                  mConnected;
+    private HybiParser mParser;
+    private boolean mConnected;
+
+    private DataCallback mDataCallback;
+    private StringCallback mStringCallback;
+    private ClosedCallback mClosedCallback;
+
+    private HappyDataInputStream stream;
```
Comment (koush) [inline src/com/codebutler/android_websockets/WebSocketClient.java]: Was there a reason to move this stream var as a member?

File: src/com/codebutler/android_websockets/WebSocketClient.java [hunk 5]
```diff
@@ -173,7 +196,7 @@ public void send(byte[] data) {
         sendFrame(mParser.frame(data));
     }
 
-    public boolean isConnected() {
```
Comment (koush) [inline src/com/codebutler/android_websockets/WebSocketClient.java:176]: Rename this back as well

**Comment by koush** [comment]
@codebutler Awesome, thanks.

File: src/com/codebutler/android_websockets/WebSocketClient.java (hunk 3 above)
Comment (vinaysshenoy) [inline src/com/codebutler/android_websockets/WebSocketClient.java:31]: Yeah, sorry about that. That's my Eclipse formatter. I've got into the habit of hitting Ctrl+Shift+F just before I save and it ends up realigning some stuff. I'll have to fix that.

Comment (koush) [inline src/com/codebutler/android_websockets/WebSocketClient.java:31]: I suspect this may have just been your Eclipse or whatever autoformatting.

File: src/com/codebutler/android_websockets/WebSocketClient.java (hunk 2 above)
Comment (vinaysshenoy) [inline src/com/codebutler/android_websockets/WebSocketClient.java]: All right. I need to get some sleep now, and I'll revert back to Listener first thing tomorrow.

File: src/com/codebutler/android_websockets/WebSocketClient.java (hunk 4 above)
Comment (vinaysshenoy) [inline src/com/codebutler/android_websockets/WebSocketClient.java]: It was actually due to the Callback methods since the attach() method in SocketIOConnection is called after the WebSocket returns. I didn't want to start parsing before the callbacks were added. Since I'm reverting back to the older Listener interface, this can go as well.

File: src/com/codebutler/android_websockets/WebSocketClient.java (hunk 5 above)
Comment (vinaysshenoy) [inline src/com/codebutler/android_websockets/WebSocketClient.java:176]: The isConnected() method is unchanged from the original. Did you mean to indicate something else?

File: src/com/codebutler/android_websockets/WebSocketClient.java [hunk 6]
```diff
@@ -173,7 +196,7 @@ public void send(byte[] data) {
...
+    public boolean isOpen() {
         return mConnected;
```
Comment (koush) [inline src/com/codebutler/android_websockets/WebSocketClient.java:177]: > The isConnected() method is unchanged from the original. Did you mean to indicate something else?

Yes, isConnected was renamed to isOpen

**Comment by vinaysshenoy** [comment]
@koush Done. All switched back to the original WebSocketClient/HybiParser implementation. Changes are limited to the SocketIO classes only.

**Comment by koush** [comment]
Perfect, thanks!
